spyder/plugins/plots/widgets/main_widget.py

```diff
@@ -0,0 +1,566 @@
+                    thumbnails_sb_actions = thumbnails_sb.actions()
+
+                    if action not in figviewer_actions:
+                        figviewer.addAction(action)
+
+                    if action not in thumbnails_sb_actions:
+                        thumbnails_sb.addAction(action)
+
+        self.zoom_disp.setEnabled(value)
+
+        # Disable zoom buttons if autofit
+        if value:
+            value = not self.get_option('auto_fit_plotting')
+            self.get_action(PlotsWidgetActions.ZoomIn).setEnabled(value)
+            self.get_action(PlotsWidgetActions.ZoomOut).setEnabled(value)
+            self.zoom_disp.setEnabled(value)
+
+    def on_option_update(self, option, value):
+        for index in range(self.count()):
+            widget = self._stack.widget(index)
+            if widget:
+                widget.setup({option: value})
+                self.update_actions()
+
+    # --- Public API:
+    # ------------------------------------------------------------------------
+    def set_current_widget(self, fig_browser):
+        """
+        Set the current figure browser widget in the stack.
+
+        Parameters
+        ----------
+        fig_browser: spyder.plugins.plots.widgets.figurebrowser.FigureBrowser
+            The widget to set.
+        """
+        self._stack.setCurrentWidget(fig_browser)
+
+    def current_widget(self):
+        """
+        Return the current figure browser widget in the stack.
+
+        Returns
+        -------
+        spyder.plugins.plots.widgets.figurebrowser.FigureBrowser
+            The current widget.
+        """
+        return self._stack.currentWidget()
+
+    def count(self):
+        """
+        Return the number of widgets in the stack.
+
+        Returns
+        -------
+        int
+            The number of widgets in the stack.
+        """
+        return self._stack.count()
+
+    def remove_widget(self, fig_browser):
+        """
+        Remove widget from stack.
+
+        Parameters
+        ----------
+        fig_browser: spyder.plugins.plots.widgets.figurebrowser.FigureBrowser
+            The figure browser widget to remove.
+        """
+        self._stack.removeWidget(fig_browser)
+
+    def add_widget(self, fig_browser):
+        """
+        Add widget to stack.
+
+        Parameters
+        ----------
+        fig_browser: spyder.plugins.plots.widgets.figurebrowser.FigureBrowser
+            The figure browser widget to add.
+        """
+        self._stack.addWidget(fig_browser)
+
+    def add_shellwidget(self, shellwidget):
+        """
+        Add a new shellwidget registered with the plots plugin.
+
+        This function registers a new FigureBrowser for browsing the figures
+        in the shell.
+
+        Parameters
+        ----------
+        shelwidget: spyder.plugins.ipyconsole.widgets.shell.ShellWidget
+            The shell widget.
+        """
+        shellwidget_id = id(shellwidget)
```

Comment:
Ok, so I can see here that it's perfectly possible to use the `shellwidget` id instead of the object itself. Since that's what we really need to emit from the IPython console, I think you should restore passing id's between the IPython console and the plugins that depend on detecting the current `shellwidget` (i.e. Plots, Variable Explorer and Help).

On a side note, this is what we've been doing since we integrated the IPython console in Spyder eight years ago (and it was probably there before to handle the interaction with the Python console too). So I don't understand why you decided to change it and I can't think of a compelling reason to do it.

Reply:
>So I don't understand why you decided to change it and I can't think of a compelling reason to do it.

The names of the signals were the same but some were passing IDs and some were passing shellwidgets. Still the shell widget is needed.

>Spyder eight years ago (and it was probably there before to handle the interaction with the Python console too)

Doing things the same way it has been done for X amount of time is rarely a good excuse to stop doing things. If that was the case this work would not have started in the first place.